### PR TITLE
Feat!: Support pre-post statements in python models at creation time

### PR DESCRIPTION
**Pull request by Themiscodes**
This update introduces the ability to enable pre- and post-statements in python models that can be executed during the table creation phase, fixes: #2866 

The SqlBased model renderers has been moved to the `_Model` class, so that the logic can be reused by python models. Also to support macro calls in the pre- and post-statements, the python environment is updated with the referenced macros (similarly to sql models).

### CODE REVIEW

File: sqlmesh/core/model/definition.py [hunk 1]
```diff
@@ -1877,6 +1852,21 @@ def create_python_model(
...
     # Find dependencies for python models by parsing code if they are not explicitly defined
     # Also remove self-references that are found
+
+    pre_statements = kwargs.get("pre_statements", None) or []
+    post_statements = kwargs.get("post_statements", None) or []
+
+    python_env.update(
+        _python_env(
+            expressions=[*pre_statements, *post_statements],
+            jinja_macro_references=None,
```
Comment (georgesittas) [inline sqlmesh/core/model/definition.py:1862]: Let's make sure jinja can be used for pre/post-statements in python models too, otherwise the functionality is unnecessarily crippled, right? This section may be useful: https://github.com/TobikoData/sqlmesh/blob/6e71e5fd60ad68062124510b4dc97dc83332facb/sqlmesh/core/model/definition.py#L1627

File: sqlmesh/core/model/definition.py [hunk 2]
```diff
@@ -1853,8 +1825,11 @@ def create_seed_model(
 def create_python_model(
     name: str,
     entrypoint: str,
+    module_path: Path,
```
Comment (georgesittas) [inline sqlmesh/core/model/definition.py:1828]: Why was this added as a positional argument? This makes the PR breaking right? Let's double check this, these API changes are a bit risky.

File: sqlmesh/core/model/decorator.py [hunk 3]
```diff
@@ -123,7 +127,7 @@ def model(
         for key in ("pre_statements", "post_statements"):
             statements = common_kwargs.get(key)
             if statements:
-                common_kwargs[key] = [exp.maybe_parse(s, dialect=dialect) for s in statements]
+                common_kwargs[key] = [parse_one(s, dialect=dialect) for s in statements]
```
Comment (georgesittas) [inline sqlmesh/core/model/decorator.py:130]: Can `statements` ever contain actual expression instances instead of strings? This will crash if this happens. Perhaps that's why we were using `exp.maybe_parse` here in the first place?

Comment (Themiscodes) [reply sqlmesh/core/model/decorator.py]: right, I didn't consider this. I've adjusted it to allow the use of actual expression instances

**Review by georgesittas — APPROVED** [review]

**Review by georgesittas — APPROVED** [review]
